File: Priority Queue/kth-largest-element-in-a-stream-703.py
# https://leetcode.com/problems/kth-largest-element-in-a-stream/

# TL;DR: 
## Using a min heap to keep track of the first k elements always and have the lowest element as the first one:
##  - in the init method we initialize the heap with all elems of nums
##      - if there are more elems in the heap than k, pop the lowest until it len(heap) == k
##  - in the add method push the new elem in the heap
##      - if the new heap length is greater thank k pop the lowest again
##      - return the first element from the heap (the currently lowest) which is the kth largest elem


# Complexities:
## Time => O(n * log(k)), where n is the len(nums) and k is the input number (size of the heap)
## Space => O(k), where k is the input number (size of the heap)

class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self.heap, self.k = nums, k
        heapq.heapify(self.heap)
        while len(self.heap) > k:
            heapq.heappop(self.heap)

    def add(self, val: int) -> int:
        heapq.heappush(self.heap, val)
        
        if len(self.heap) > self.k:
            heapq.heappop(self.heap)
        
        return self.heap[0]

File: Priority Queue/kth-largest-element-in-a-stream-703.ts
// https://leetcode.com/problems/kth-largest-element-in-a-stream/

// TL;DR:
// Using a min heap to keep track of the first k elements always and have the lowest element as the first one:
//  - in the constructor iterate through nums and push each element to the heap while popping the smallest if we overflow over k
//  - in the add method push the new elem in the heap and if we overflow over k pop the lowest again, then return the first element from the heap (the currently lowest)

// Complexities:
// Time => O(n * log(k)), where n is the length of nums and k is the input number (size of the heap)
// Space => O(k), where k is the input number (size of the heap)

import { MinPriorityQueue } from "@datastructures-js/priority-queue";

class KthLargest {
	private readonly minHeap: MinPriorityQueue<number>;
	private readonly k: number;

	constructor(k: number, nums: number[]) {
		this.k = k;

		this.minHeap = new MinPriorityQueue();
		for (const num of nums) {
			this.minHeap.enqueue(num);
			if (this.minHeap.size() > k) {
				this.minHeap.dequeue();
			}
		}
	}

	add(val: number): number {
		this.minHeap.enqueue(val);
		if (this.minHeap.size() > this.k) {
			this.minHeap.dequeue();
		}
		return this.minHeap.front();
	}
}
